webapp/login/views.py

```diff
@@ -96,6 +105,72 @@ def after_login(resp):
     return flask.redirect(open_id.get_next_url())
 
 
+@login.route("/login-beta", methods=["GET"])
+@csrf.exempt
+def login_candid():
+    # Get a bakery v2 macaroon from the publisher API to be discharged
+    # and save it in the session
+    flask.session["publisher-macaroon"] = publisher_api.get_macaroon(
+        authentication.PERMISSIONS
+    )
+
+    login_url = candid.get_login_url(
+        macaroon=flask.session["publisher-macaroon"],
+        callback_url=flask.url_for("login.login_callback", _external=True),
+        state=generate_csrf(),
+    )
+
+    # Next URL to redirect the user after the login
+    next_url = flask.request.args.get("next")
+
+    if next_url:
+        if not next_url.startswith(
+            flask.request.url_root
+        ) or next_url.startswith("/"):
+            return flask.abort(400)
+        flask.session["next_url"] = next_url
+
+    return flask.redirect(login_url, 302)
+
+
+@login.route("/login/callback")
+def login_callback():
+    code = flask.request.args["code"]
+    state = flask.request.args["state"]
+
+    # Avoid CSRF attacks
+    validate_csrf(state)
+
+    discharged_token = candid.discharge_token(code)
+    candid_macaroon = candid.discharge_macaroon(
+        flask.session["publisher-macaroon"], discharged_token
+    )
+
+    # Store bakery authentication
+    flask.session["macaroons"] = candid.get_serialized_bakery_macaroon(
+        flask.session["publisher-macaroon"], candid_macaroon
+    )
+
+    try:
+        publisher = publisher_api.whoami(flask.session)
+    except StoreApiResponseErrorList as api_response_error_list:
+        return _handle_error_list(api_response_error_list.errors)
+    except (StoreApiError, ApiError) as api_error:
+        return _handle_error(api_error)
+
+    # whoami is the only Dashboard API endpoint
+    # that support Candid.
+    return publisher
+
+    # return flask.redirect(
```

Comment:
Forgot to remove?

Reply:
Not really, the API doesn't do anything yet. I thought it will be better to return the publisher details after the login until the API is ready.

Reply:
I mean I can remove that commented code for now and add it back later.

Reply:
so it seems I can login but I keep getting the `Developer Programme Agreement` page.

Reply:
nevermind it seems my old account was in a weird state.